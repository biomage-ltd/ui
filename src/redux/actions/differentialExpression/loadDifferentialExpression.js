import {
  DIFF_EXPR_LOADING, DIFF_EXPR_LOADED, DIFF_EXPR_ERROR,
} from '../../actionTypes/differentialExpression';

import { fetchCachedWork } from '../../../utils/cacheRequest';

const getCellSetName = (name) => (name?.split('/')[1] || name);

const REQUEST_TIMEOUT = 60;
const loadDifferentialExpression = (
  experimentId, cellSets, comparisonType, tableState,
) => async (dispatch, getState) => {
  const {
    startDate: pipelineStartDate,
  } = getState().experimentSettings.backendStatus.status.pipeline;

  dispatch({
    type: DIFF_EXPR_LOADING,
    payload: {
      experimentId,
    },
  });

  const body = {
    name: 'DifferentialExpression',
    experimentId,
    cellSet: getCellSetName(cellSets.cellSet),
    compareWith: getCellSetName(cellSets.compareWith),
    basis: getCellSetName(cellSets.basis),
  };

  let pagination = {};

  if (tableState) {
    const currentPageSize = tableState.pagination.pageSize;

    pagination = {
      orderBy: tableState.sorter.field,
      orderDirection: (tableState.sorter.order === 'ascend') ? 'ASC' : 'DESC',
      offset: ((tableState.pagination.current - 1) * currentPageSize),
      limit: currentPageSize,
      responseKey: 0,
    };

    if (tableState.geneNamesFilter) {
      pagination.filters = [{
        columnName: 'gene_names',
        type: 'text',
        expression: tableState.geneNamesFilter,
      }];
    }

    pagination = { pagination };
  }

  try {
    const data = await fetchCachedWork(
      experimentId, REQUEST_TIMEOUT, body, pipelineStartDate, pagination,
    );

    let { total } = data;
    const { rows } = data;

    if (!total && !Object.keys(pagination).length) {
      total = rows.length;
    }

    return dispatch({
      type: DIFF_EXPR_LOADED,
      payload: {
        experimentId,
        data: rows,
        cellSets,
        total,
        comparisonType,
      },
    });
  } catch (error) {
    dispatch({
      type: DIFF_EXPR_ERROR,
      payload: {
        experimentId,
        error,
      },
    });
  }
};

export default loadDifferentialExpression;
