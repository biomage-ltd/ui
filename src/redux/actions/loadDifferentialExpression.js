import {
  DIFF_EXPR_LOADING, DIFF_EXPR_LOADED, DIFF_EXPR_ERROR,
} from '../actionTypes/differentialExpression';

import sendWork from '../../utils/sendWork';

const REQUEST_TIMEOUT = 60;
const loadDifferentialExpression = (
  experimentId, cellSets, tableState,
) => async (dispatch) => {
  dispatch({
    type: DIFF_EXPR_LOADING,
    payload: {
      experimentId,
    },
  });

  const body = {
    name: 'DifferentialExpression',
    ...cellSets,
  };

  let pagination = {};

  if (tableState) {
    const currentPageSize = tableState.pagination.pageSize;

    pagination = {
      orderBy: tableState.sorter.field,
      orderDirection: (tableState.sorter.order === 'ascend') ? 'ASC' : 'DESC',
      offset: ((tableState.pagination.current - 1) * currentPageSize),
      limit: currentPageSize,
      responseKey: 0,
    };

    if (tableState.geneNamesFilter) {
      pagination.filters = [{
        columnName: 'gene_names',
        type: 'text',
        expression: tableState.geneNamesFilter,
      }];
    }

    pagination = { pagination };
  }

  try {
    const response = await sendWork(
      experimentId, REQUEST_TIMEOUT, body, pagination,
    );
    const data = JSON.parse(response.results[0].body);
    let { total } = data;
    const { rows } = data;

    if (!total && !Object.keys(pagination).length) {
      total = rows.length;
    }
    return dispatch({
      type: DIFF_EXPR_LOADED,
      payload: {
        experimentId,
        data: rows,
        cellSets,
        total,
      },
    });
  } catch (error) {
    console.error(error);
    dispatch({
      type: DIFF_EXPR_ERROR,
      payload: {
        experimentId,
        error: "Couldn't fetch differential expression results data.",
      },
    });
  }
};

export default loadDifferentialExpression;
