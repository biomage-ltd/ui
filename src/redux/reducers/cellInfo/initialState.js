const initialState = {
  focus: {
    store: 'cellSets',
    key: 'louvain',
  },
  groupedTrack: 'louvain',
  selectedTracks: ['louvain'],
};

export default initialState;
