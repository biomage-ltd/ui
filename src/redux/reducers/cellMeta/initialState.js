const mitochondrialContent = {
  loading: true,
  error: false,
  data: [],
};

const doubletScores = {
  loading: true,
  error: false,
  data: [],
};

const initialState = {
  mitochondrialContent,
  doubletScores,
};

export default initialState;
