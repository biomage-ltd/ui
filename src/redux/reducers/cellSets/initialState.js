const initialState = {
  properties: {},
  hierarchy: [],
  selected: [],
  loading: true,
  error: false,
  updatingClustering: false,
  hidden: new Set(),
};

export default initialState;
