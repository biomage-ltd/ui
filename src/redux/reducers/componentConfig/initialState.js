import {
  legendBaseState,
  dimensionsBaseState,
  axesBaseState,
  titleBaseState,
  fontStyleBaseState,
  colourBaseState,
  markerBaseState,
  labelBaseState,
} from './baseStylesState';

// PLOTS & TABLES - Categorical Embedding
const embeddingCategoricalInitialConfig = {
  spec: '1.0.0',
  legend: {
    ...legendBaseState,
    position: 'top',
  },
  dimensions: {
    ...dimensionsBaseState,
    width: 700,
    height: 550,
  },
  axes: {
    ...axesBaseState,
    xAxisText: 'UMAP 1',
    yAxisText: 'UMAP 2',
    offset: 10,
  },
  title: {
    ...titleBaseState,
    fontSize: 20,
  },
  fontStyle: fontStyleBaseState,
  colour: colourBaseState,
  marker: markerBaseState,
  label: labelBaseState,
  selectedCellSet: 'louvain',
  selectedSample: 'All',
};

// PLOTS & TABLES - Continuous Embedding
const embeddingContinuousInitialConfig = {
  spec: '1.0.0',
  legend: legendBaseState,
  dimensions: {
    ...dimensionsBaseState,
    width: 700,
    height: 550,
  },
  axes: {
    ...axesBaseState,
    xAxisText: 'UMAP 1',
    yAxisText: 'UMAP 2',
    offset: 10,
  },
  title: {
    ...titleBaseState,
    dx: 0,
    fontSize: 20,
  },
  fontStyle: fontStyleBaseState,
  colour: colourBaseState,
  marker: markerBaseState,
  label: labelBaseState,
  logEquation: 'datum.expression*1',
  shownGene: 'notSelected',
  selectedSample: 'All',
};

// PLOTS & TABLES - Heatmap
const heatmapInitialConfig = {
  spec: '1.0.0',
  legend: {
    ...legendBaseState,
    show: true,
    position: 'horizontal',
  },
  dimensions: {
    ...dimensionsBaseState,
    width: 500,
    height: 500,
  },
  title: {
    ...titleBaseState,
    fontSize: 20,
    dx: 0,
  },
  fontStyle: fontStyleBaseState,
  colour: colourBaseState,
  marker: markerBaseState,
  label: labelBaseState,
  selectedGenes: [],
  selectedCellSet: 'louvain',
  labelColour: 'transparent',
};

// PLOTS & TABLES - Volcano plot
const volcanoInitialConfig = {
  spec: '1.0.0',
  legend: legendBaseState,
  dimensions: dimensionsBaseState,
  marker: {
    ...markerBaseState,
    size: 32,
  },
  axes: {
    ...axesBaseState,
    xAxisText: 'log2 fold change',
    yAxisText: ' - log10(adj.p - value)',
    gridOpacity: 5,
    offset: 10,
  },
  title: titleBaseState,
  fontStyle: fontStyleBaseState,
  colour: colourBaseState,
  label: labelBaseState,
  noDifferenceColor: '#aaaaaa',
  significantUpregulatedColor: '#0000ffaa',
  significantDownregulatedColor: '#ff0000',
  notSignificantDownregulatedColor: '#aaaaaa',
  notSignificantUpregulatedColor: '#aaaaaa',
  significantChangeDirectionUnknownColor: '#aaaaaa',

  // `null` automatically scales to range. This is a problem
  // because our DE is bad right now, so it throws off the
  // range to extreme values. TODO: set this back when we have
  // good DE
  // logFoldChangeDomain: null,

  logFoldChangeDomain: 20,
  maxNegativeLogpValueDomain: null,
  negLogpValueThreshold: 4,
  logFoldChangeThreshold: 1,
  logFoldChangeTickCount: 5,
  negativeLogpValueTickCount: 5,
  downsampleRatio: 0,
  showLogFoldChangeThresholdGuides: true,
  showpvalueThresholdGuides: true,
  thresholdGuideWidth: 1,
  logFoldChangeThresholdColor: '#ff0000',
  pvalueThresholdColor: '#ff0000',
  textThresholdValue: 240,
  strokeOpa: 1,
  strokeCol: '#000000',
};

// PLOTS & TABLES - Frequency
const frequencyInitialConfig = {
  spec: '1.0.0',
  frequencyType: 'proportional',
  proportionGrouping: '',
  legend: {
    ...legendBaseState,
    title: 'Cell Set',
    position: 'top',
    offset: 40,
  },
  label: labelBaseState,
  dimensions: dimensionsBaseState,
  marker: markerBaseState,
  colour: colourBaseState,
  title: titleBaseState,
  axes: {
    ...axesBaseState,
    xAxisText: 'Sample',
    yAxisText: 'Proportion',
    offset: 10,
  },
  fontStyle: fontStyleBaseState,
  xAxisGrouping: '',
  axisTitlesize: 13,
  geneexpLegendloc: '',
};

// EMBEDDING PREVIEW - Coloured by sample
const embeddingPreviewBySampleInitialConfig = {
  spec: '1.0.0',
  legend: {
    ...legendBaseState,
    position: 'top',
  },
  dimensions: {
    ...dimensionsBaseState,
    width: 700,
    height: 550,
  },
  axes: {
    ...axesBaseState,
    xAxisText: 'UMAP 1',
    yAxisText: 'UMAP 2',
    offset: 10,
  },
  title: {
    ...titleBaseState,
    fontSize: 20,
  },
  fontStyle: fontStyleBaseState,
  colour: colourBaseState,
  marker: markerBaseState,
  label: labelBaseState,
  selectedCellSet: 'louvain',
  selectedSample: 'All',
};

// EMBEDDING PREVIEW - Coloured by cell sets
const embeddingPreviewByCellSetsInitialConfig = {
  spec: '1.0.0',
  legend: {
    ...legendBaseState,
    position: 'top',
  },
  dimensions: {
    ...dimensionsBaseState,
    width: 700,
    height: 550,
  },
  axes: {
    ...axesBaseState,
    xAxisText: 'UMAP 1',
    yAxisText: 'UMAP 2',
    offset: 10,
  },
  title: {
    ...titleBaseState,
    fontSize: 20,
  },
  fontStyle: fontStyleBaseState,
  colour: colourBaseState,
  marker: markerBaseState,
  label: labelBaseState,
  selectedCellSet: 'louvain',
  selectedSample: 'All',
};

// EMBEDDING PREVIEW - Config for fraction of mitochondrial reads
const embeddingPreviewMitochondrialContentInitialConfig = {
  spec: '1.0.0',
  legend: legendBaseState,
  dimensions: {
    ...dimensionsBaseState,
    width: 700,
    height: 550,
  },
  axes: {
    ...axesBaseState,
    xAxisText: 'UMAP 1',
    yAxisText: 'UMAP 2',
    offset: 10,
  },
  title: {
    ...titleBaseState,
    dx: 0,
    fontSize: 20,
  },
  fontStyle: fontStyleBaseState,
  colour: colourBaseState,
  marker: markerBaseState,
  label: labelBaseState,
  selectedSample: 'All',
};

// EMBEDDING PREVIEW - Config for doublet score
const embeddingPreviewDoubletScoreInitialConfig = {
  spec: '1.0.0',
  legend: legendBaseState,
  dimensions: {
    ...dimensionsBaseState,
    width: 700,
    height: 550,
  },
  axes: {
    ...axesBaseState,
    xAxisText: 'UMAP 1',
    yAxisText: 'UMAP 2',
    offset: 10,
  },
  title: {
    ...titleBaseState,
    dx: 0,
    fontSize: 20,
  },
  fontStyle: fontStyleBaseState,
  colour: colourBaseState,
  marker: markerBaseState,
  label: labelBaseState,
  selectedSample: 'All',
};

const interactiveHeatmapInitialConfig = {
  selectedTracks: ['louvain'],
  groupedTracks: ['sample', 'louvain'],
  expressionValue: 'raw',
  legendIsVisible: true,
};

// CELL SIZE DISTRIBUTION - Cell Size Distribution Histogram
const cellSizeDistributionHistogram = {
  spec: '1.0.0',
  legend: {
    ...legendBaseState,
    position: 'top-left',
  },
  dimensions: {
    ...dimensionsBaseState,
    width: 530,
    height: 400,
  },
  axes: {
    ...axesBaseState,
    xAxisText: '#UMIs in cell',
    yAxisText: '#UMIs * #Cells',
  },
  title: {
    ...titleBaseState,
    fontSize: 20,
  },
  fontStyle: fontStyleBaseState,
  colour: colourBaseState,
  marker: markerBaseState,
  label: labelBaseState,
  minCellSize: 10800,
  binStep: 200,
};

// CELL SIZE DISTRIBUTION - Cell Size Distribution Knee Plot
const cellSizeDistributionKneePlot = {
  spec: '1.0.0',
  legend: {
    ...legendBaseState,
    position: 'top-left',
  },
  dimensions: {
    ...dimensionsBaseState,
    width: 530,
    height: 400,
  },
  axes: {
    ...axesBaseState,
    xAxisText: 'Cell rank',
    yAxisText: '#UMIs in cell',
  },
  title: {
    ...titleBaseState,
    fontSize: 20,
  },
  fontStyle: fontStyleBaseState,
  colour: colourBaseState,
  marker: markerBaseState,
  label: labelBaseState,
  minCellSize: 990,
};

// MITOCHONDRIAL CONTENT - Mitochondrial Fraction Histogram
const mitochondrialFractionHistogram = {
  spec: '1.0.0',
  legend: {
    ...legendBaseState,
    position: 'top-right',
  },
  dimensions: {
    ...dimensionsBaseState,
    width: 530,
    height: 400,
  },
  axes: {
    ...axesBaseState,
    xAxisText: 'Fraction of mitochondrial reads',
    yAxisText: 'Fraction of cells',
  },
  title: {
    ...titleBaseState,
    fontSize: 20,
    dx: 0,
  },
  fontStyle: fontStyleBaseState,
  colour: colourBaseState,
  marker: markerBaseState,
  label: labelBaseState,
  binStep: 0.05,
  maxFraction: 0.1,
};

// MITOCHONDRIAL CONTENT - Mitochondrial Fraction Log Histogram
const mitochondrialFractionLogHistogram = {
  spec: '1.0.0',
  legend: {
    ...legendBaseState,
    position: 'top-right',
  },
  dimensions: {
    ...dimensionsBaseState,
    width: 530,
    height: 400,
  },
  axes: {
    ...axesBaseState,
    xAxisText: 'log10(#UMIs in cell)',
    yAxisText: 'Fraction of mitochondrial reads',
  },
  title: {
    ...titleBaseState,
    fontSize: 20,
    dx: 0,
  },
  fontStyle: fontStyleBaseState,
  colour: colourBaseState,
  marker: markerBaseState,
  label: labelBaseState,
  binStep: 0.05,
  maxFraction: 3.5,
};

// CLASSIFIER - Classifier Empty Drops
const classifierEmptyDropsPlot = {
  spec: '1.0.0',
  legend: {
    ...legendBaseState,
    enabled: false,
  },
  dimensions: {
    ...dimensionsBaseState,
    width: 630,
    height: 500,
  },
  axes: {
    ...axesBaseState,
    xAxisText: 'log10[ cell size (UMIs) ]',
    yAxisText: 'classifier prob',
    gridOpacity: 10,
  },
  title: {
    ...titleBaseState,
    fontSize: 20,
  },
  fontStyle: fontStyleBaseState,
  colour: colourBaseState,
  marker: markerBaseState,
  label: labelBaseState,
  minProbability: 0.82,
  bandwidth: -1,
};

// GENES VS UMIS - Features vs UMIs
const featuresVsUMIsHistogram = {
  spec: '1.0.0',
  dimensions: {
    ...dimensionsBaseState,
    width: 630,
    height: 500,
  },
  axes: {
    ...axesBaseState,
    xAxisText: 'log10 [molecules]',
    yAxisText: 'Frequency',
    gridOpacity: 10,
  },
  title: {
    ...titleBaseState,
    fontSize: 20,
  },
  fontStyle: fontStyleBaseState,
  colour: colourBaseState,
  label: labelBaseState,
  lowCutoff: 4.8,
  upCutoff: 2.1,
  binStep: 0.05,
};

// GENES VS UMIS - Features vs UMIs
const featuresVsUMIsScatterplot = {
  spec: '1.0.0',
  dimensions: {
    ...dimensionsBaseState,
    width: 630,
    height: 500,
  },
  axes: {
    ...axesBaseState,
    xAxisText: 'log10 [molecule counts]',
    yAxisText: 'log10 [gene counts]',
    gridOpacity: 10,
  },
  title: {
    ...titleBaseState,
    fontSize: 20,
  },
  fontStyle: fontStyleBaseState,
  colour: colourBaseState,
  label: labelBaseState,
  lowCutoff: 4.8,
  upCutoff: 2.1,
};

// DOUBLE FILTER - Doublet filter histogram
const doubletScoreHistogram = {
  spec: '1.0.0',
  legend: {
    ...legendBaseState,
    position: 'top-right',
  },
  dimensions: {
    ...dimensionsBaseState,
    width: 530,
    height: 400,
  },
  axes: {
    ...axesBaseState,
    xAxisText: 'Probability of being a doublet',
    yAxisText: 'Frequency',
  },
  title: {
    ...titleBaseState,
    fontSize: 20,
  },
  fontStyle: fontStyleBaseState,
  colour: colourBaseState,
  marker: markerBaseState,
  label: labelBaseState,
  binStep: 0.05,
  probThreshold: 0.2,
};

// DATA INTEGRATION - Embedding by Samples
const dataIntegrationEmbeddingInitialConfig = {
  spec: '1.0.0',
  legend: {
    ...legendBaseState,
    position: 'top',
  },
  dimensions: {
    ...dimensionsBaseState,
    width: 700,
    height: 550,
  },
  axes: {
    ...axesBaseState,
    xAxisText: 'UMAP 1',
    yAxisText: 'UMAP 2',
    offset: 10,
  },
  title: {
    ...titleBaseState,
    fontSize: 20,
  },
  fontStyle: fontStyleBaseState,
  colour: colourBaseState,
  marker: markerBaseState,
  label: labelBaseState,
  selectedCellSet: 'louvain',
  selectedSample: 'All',
};

// DATA INTEGRATION - Frequency
const dataIntegrationFrequencyInitialConfig = {
  ...frequencyInitialConfig,
  xAxisGrouping: 'louvain',
  proportionGrouping: 'sample',
  legend: {
    ...legendBaseState,
    title: 'Sample',
    position: 'top',
    offset: 10,
  },
  dimensions: {
    ...dimensionsBaseState,
    width: 700,
    height: 550,
  },

  axes: {
    ...axesBaseState,
    xAxisText: 'Louvain clusters',
    yAxisText: 'Proportion',
    offset: 10,
  },
  fontStyle: fontStyleBaseState,
};

// DATA INTEGRATION - Elbow
const dataIntegrationElbowPlotInitialConfig = {
  legend: {
    ...legendBaseState,
    position: 'top',
  },
  label: { ...labelBaseState },
  dimensions: {
    ...dimensionsBaseState,
    width: 700,
    height: 550,
  },
  marker: { ...markerBaseState },
  fontStyle: fontStyleBaseState,
  axes: {
    ...axesBaseState,
    xAxisText: 'Principal Components',
    yAxisText: 'Proportion of Variance Explained',
    titleFont: 'sans-serif',
    labelFont: 'sans-serif',
    labelFontSize: 13,
    offset: 0,
    gridOpacity: 10,
  },
  colour: {
    toggleInvert: '#FFFFFF',
  },
  title: {
    ...titleBaseState,
    font: 'sans-serif',
    fontSize: 12,
  },
  signals: [
    {
      name: 'interpolate',
      value: 'linear',
      bind: {
        input: 'select',
        options: [
          'basis',
          'cardinal',
          'catmull-rom',
          'linear',
          'monotone',
          'natural',
          'step',
          'step-after',
          'step-before',
        ],
      },
    },
  ],
};

const initialPlotConfigStates = {
  cellSizeDistributionHistogram,
  cellSizeDistributionKneePlot,
  mitochondrialFractionHistogram,
  mitochondrialFractionLogHistogram,
  classifierEmptyDropsPlot,
  featuresVsUMIsHistogram,
  featuresVsUMIsScatterplot,
  doubletScoreHistogram,
  embeddingCategorical: embeddingCategoricalInitialConfig,
  embeddingContinuous: embeddingContinuousInitialConfig,
  heatmap: heatmapInitialConfig,
  volcano: volcanoInitialConfig,
  frequency: frequencyInitialConfig,
  embeddingPreviewBySample: embeddingPreviewBySampleInitialConfig,
  embeddingPreviewByCellSets: embeddingPreviewByCellSetsInitialConfig,
  embeddingPreviewMitochondrialContent: embeddingPreviewMitochondrialContentInitialConfig,
  embeddingPreviewDoubletScore: embeddingPreviewDoubletScoreInitialConfig,
  dataIntegrationEmbedding: dataIntegrationEmbeddingInitialConfig,
  dataIntegrationFrequency: dataIntegrationFrequencyInitialConfig,
  dataIntegrationElbow: dataIntegrationElbowPlotInitialConfig,
};

const initialComponentConfigStates = {
  interactiveHeatmap: interactiveHeatmapInitialConfig,
};
export { initialPlotConfigStates, initialComponentConfigStates };

const initialState = {};
export default initialState;
