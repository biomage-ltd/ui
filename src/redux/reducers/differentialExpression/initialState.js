const initialState = {
  properties: {
    data: [],
    cellSets: {},
    loading: false,
    error: false,
    total: 0,
  },
  comparison: {
    type: 'between',
    group: {
      cellSet: null,
      compareWith: null,
      basis: null,
    },
  },
};

export default initialState;
