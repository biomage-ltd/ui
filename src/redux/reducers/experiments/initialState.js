const experimentTemplate = {
  projectUuid: null,
  name: null,
  description: 'Analysis description',
  id: null,
  createdDate: null,
  lastViewed: null,
};

const initialState = {
  meta: {
    loading: false,
    error: false,
    saving: false,
  },
  ids: [],
};

export default initialState;
export { experimentTemplate };
