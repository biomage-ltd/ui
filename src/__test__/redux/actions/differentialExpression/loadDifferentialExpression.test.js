import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import loadDifferentialExpression from '../../../../redux/actions/differentialExpression/loadDifferentialExpression';

import initialState from '../../../../redux/reducers/differentialExpression/initialState';
import { fetchCachedWork } from '../../../../utils/cacheRequest';

import {
  DIFF_EXPR_LOADING, DIFF_EXPR_LOADED, DIFF_EXPR_ERROR,
} from '../../../../redux/actionTypes/differentialExpression';

jest.mock('localforage');
jest.mock('../../../../utils/cacheRequest');

const mockStore = configureStore([thunk]);

const defaultTableState = {
  geneNamesFilter: null,
  pagination: {
    current: 1, pageSize: 50, showSizeChanger: true, total: 0,
  },
  sorter: { field: 'p_val_adj', columnKey: 'p_val_adj', order: 'ascend' },
};

const backendStatus = {
  status: {
    pipeline: {
      startDate: '2021-01-01T00:00:00',
    },
  },
};

describe('loadDifferentialExpression action', () => {
  const experimentId = '1234';
  const cellSets = {
    cellSet: 'louvain/louvain-0',
    compareWith: 'louvain/louvain-1',
    basis: 'condition/condition-control',
  };
  const comparisonType = 'within';

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('Dispatches appropriately on failure', async () => {
    const store = mockStore({
      differentialExpression: {
        ...initialState,
      },
      experimentSettings: {
        backendStatus,
      },
    });
    fetchCachedWork.mockImplementationOnce(() => new Promise((resolve, reject) => reject(new Error('random error!'))));

    await store.dispatch(
      loadDifferentialExpression(experimentId, cellSets, comparisonType, defaultTableState),
    );

    const loadingAction = store.getActions()[0];
    expect(loadingAction.type).toEqual(DIFF_EXPR_LOADING);

    expect(fetchCachedWork).toHaveBeenCalledTimes(1);
    expect(fetchCachedWork).toHaveBeenCalledWith('1234', 60,
      {
        cellSet: 'louvain-0',
        compareWith: 'louvain-1',
        basis: 'condition-control',
        name: 'DifferentialExpression',
        experimentId: '1234',
      },
      backendStatus.status,
      {
        pagination: {
          limit: 50, offset: 0, orderBy: 'p_val_adj', orderDirection: 'ASC', responseKey: 0,
        },
      });
    expect(loadingAction).toMatchSnapshot();

    const loadedAction = store.getActions()[1];
    expect(loadedAction.type).toEqual(DIFF_EXPR_ERROR);
    expect(loadedAction).toMatchSnapshot();
  });

  it('Dispatches appropriately on success', async () => {
    const store = mockStore({
      differentialExpression: {
        ...initialState,
      },
      experimentSettings: {
        backendStatus,
      },
    });

    fetchCachedWork.mockImplementationOnce(() => {
      const resolveWith = {
        rows: [
          {
            p_val: 1.16, p_val_adj: 2.16, avg_log2FC: 3.45, gene_names: 'A',
          },
          {
            p_val: 2.16, p_val_adj: 3.16, avg_log2FC: 4.45, gene_names: 'B',
          },
          {
            p_val: 3.16, p_val_adj: 4.16, avg_log2FC: 5.45, gene_names: 'C',
          },
          {
            p_val: 4.16, p_val_adj: 5.16, avg_log2FC: 6.45, gene_names: 'D',
          },
          {
            p_val: 5.16, p_val_adj: 6.16, avg_log2FC: 7.45, gene_names: 'E',
          },
          {
            p_val: 6.16, p_val_adj: 7.16, avg_log2FC: 8.45, gene_names: 'F',
          },
        ],
      };

      return new Promise((resolve) => resolve(resolveWith));
    });

    await store.dispatch(
      loadDifferentialExpression(experimentId, cellSets, comparisonType, defaultTableState),
    );

    const loadingAction = store.getActions()[0];
    expect(loadingAction.type).toEqual(DIFF_EXPR_LOADING);
    expect(loadingAction).toMatchSnapshot();

    expect(fetchCachedWork).toHaveBeenCalledTimes(1);
    expect(fetchCachedWork).toHaveBeenCalledWith('1234', 60,
      {
        cellSet: 'louvain-0',
        compareWith: 'louvain-1',
        basis: 'condition-control',
        name: 'DifferentialExpression',
        experimentId: '1234',
      },
      backendStatus.status,
      {
        pagination: {
          limit: 50, offset: 0, orderBy: 'p_val_adj', orderDirection: 'ASC', responseKey: 0,
        },
      });

    const loadedAction = store.getActions()[1];
    expect(loadedAction.type).toEqual(DIFF_EXPR_LOADED);
    expect(loadedAction).toMatchSnapshot();
  });
});
