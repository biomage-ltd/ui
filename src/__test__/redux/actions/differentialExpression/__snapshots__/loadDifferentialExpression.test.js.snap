// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loadDifferentialExpression action Dispatches appropriately on failure 1`] = `
Object {
  "payload": Object {
    "experimentId": "1234",
  },
  "type": "diffExpr/loading",
}
`;

exports[`loadDifferentialExpression action Dispatches appropriately on failure 2`] = `
Object {
  "payload": Object {
    "error": "Couldn't fetch differential expression results data.",
    "experimentId": "1234",
  },
  "type": "diffExpr/error",
}
`;

exports[`loadDifferentialExpression action Dispatches appropriately on success 1`] = `
Object {
  "payload": Object {
    "experimentId": "1234",
  },
  "type": "diffExpr/loading",
}
`;

exports[`loadDifferentialExpression action Dispatches appropriately on success 2`] = `
Object {
  "payload": Object {
    "cellSets": Object {
      "basis": "condition/condition-control",
      "cellSet": "louvain/louvain-0",
      "compareWith": "louvain/louvain-1",
    },
    "comparisonType": "within",
    "data": Array [
      Object {
        "gene_names": "A",
        "avg_log2FC": 3.45,
        "p_val": 1.16,
        "p_val_adj": 2.16,
      },
      Object {
        "gene_names": "B",
        "avg_log2FC": 4.45,
        "p_val": 2.16,
        "p_val_adj": 3.16,
      },
      Object {
        "gene_names": "C",
        "avg_log2FC": 5.45,
        "p_val": 3.16,
        "p_val_adj": 4.16,
      },
      Object {
        "gene_names": "D",
        "avg_log2FC": 6.45,
        "p_val": 4.16,
        "p_val_adj": 5.16,
      },
      Object {
        "gene_names": "E",
        "avg_log2FC": 7.45,
        "p_val": 5.16,
        "p_val_adj": 6.16,
      },
      Object {
        "gene_names": "F",
        "avg_log2FC": 8.45,
        "p_val": 6.16,
        "p_val_adj": 7.16,
      },
    ],
    "experimentId": "1234",
    "total": undefined,
  },
  "type": "diffExpr/loaded",
}
`;
